File: Java/org/oobench/messages/MessagesBenchmark.java
package org.oobench.messages;

import org.oobench.common.*;

public abstract class MessagesBenchmark extends AbstractBenchmark {

    public int getMajorNumber() {
        return 5;
    }

    public int getMinorNumber() {
        return 1;
    }

    public abstract void invoke(int count);

    public void test(int count, String typeOfMessage) {
        count = getCountWithDefault(count);

        for (int c = 0; c < 10; c++) {
            if (c == 9) {
                beginAction(1, "invoke", count, typeOfMessage);
            }
            invoke(count);
            if (c == 9) {
                endAction();
            }
        }
    }

    public static void testMessages(Class theClass, int count, 
            String typeOfMessage, String[] args) {
        try {
            MessagesBenchmark bench = 
                    (MessagesBenchmark)theClass.newInstance();
            bench.setArguments(args);
            bench.test(count, typeOfMessage);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: C++/Messages/MessagesBenchmark.cpp
#include "AbstractBenchmark.cpp"

namespace oobench {

    class MessagesBenchmark : public AbstractBenchmark {
    public:
        MessagesBenchmark(int argc, const char** argv) 
                : AbstractBenchmark(argc, argv) { 
        }

        virtual int getMajorNumber() const {
            return 5;
        }

        virtual int getMinorNumber() const {
            return 1;
        }

        virtual void invoke(unsigned long count) = 0;

        void test(unsigned long items, const char* description) {
            unsigned long count = getCountWithDefault(items);
            beginAction(1, "invoke", count, description);
            invoke(count);
            endAction();
        }
    };
}
